pkg/codegen/schema/loader.go

```diff
@@ -210,3 +250,106 @@ func LoadPackageReference(loader Loader, pkg string, version *semver.Version) (P
 	}
 	return p.Reference(), nil
 }
+
+func (l *pluginLoader) loadSchemaBytes(pkg string, version *semver.Version) ([]byte, *semver.Version, error) {
+	pluginInfo, err := workspace.GetPluginInfo(workspace.ResourcePlugin, pkg, nil, false /* skipMetadata */)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if version == nil {
+		version = pluginInfo.Version
+	}
+
+	cachedSchemaPath := filepath.Join(pluginInfo.Path, fmt.Sprintf("schema-%v.json", pkg))
+
+	schemaBytes, ok := l.loadCachedSchemaBytes(pkg, cachedSchemaPath, pluginInfo.InstallTime)
+	if ok {
+		return schemaBytes, nil, nil
+	}
+
+	schemaBytes, provider, err := l.loadPluginSchemaBytes(pkg, version)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Error loading schema from plugin: %w", err)
+	}
+
+	// TODO(friel) disallow caching when the plugin isn't installed in a well known location, i.e.: in the plugin path
+	err = os.WriteFile(cachedSchemaPath, schemaBytes, 0600)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Error writing schema from plugin to cache: %w", err)
+	}
+
+	if version == nil {
+		info, err := provider.GetPluginInfo()
+		if err != nil {
+			// Nonfatal
+		}
+		version = info.Version
+	}
+
+	return schemaBytes, version, nil
+}
+
+func (l *pluginLoader) loadPluginSchemaBytes(pkg string, version *semver.Version) ([]byte, plugin.Provider, error) {
+	if err := l.ensurePlugin(pkg, version); err != nil {
+		return nil, nil, err
+	}
+
+	provider, err := l.host.Provider(tokens.Package(pkg), version)
+	if err != nil {
+		return nil, nil, err
+	}
+	contract.Assert(provider != nil)
+
+	schemaFormatVersion := 0
+	schemaBytes, err := provider.GetSchema(schemaFormatVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return schemaBytes, provider, nil
+}
+
+func (l *pluginLoader) loadCachedSchemaBytes(pkg string, path string, pluginCreatedAt time.Time) ([]byte, bool) {
+	if l.cacheOptions.disableFileCache {
+		return nil, false
+	}
+
+	if pluginCreatedAt.IsZero() {
+		return nil, false
+	}
+
+	schemaFile, err := os.OpenFile(path, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, false
+	}
+
+	stat, err := schemaFile.Stat()
+	if err != nil {
+		return nil, false
+	}
+	cachedAt := stat.ModTime()
+
+	if cachedAt.Before(pluginCreatedAt) {
```

Comment:
I am a little nervous about using file times here, but I don't have a better option besides e.g. checksumming the provider.

Reply:
I've changed this comparison to use an expiration "`SchemaTime`" provided by the host on PluginInfo. For `deploytest`, the time is zero, for file based plugins it's the modified-at timestamp.

This is reliable, as building a new binary or touching the script that runs it - component provider implemented in another language - invalidates the cache.

I found this a bit easier than doing a hash or checksum.